streams/src/main/java/org/apache/kafka/streams/processor/assignment/TaskAssignmentUtils.java

```diff
@@ -407,4 +543,345 @@ private static boolean hasValidRackInformation(final TaskInfo task,
+        final Map<ProcessId, KafkaStreamsState> streamStates = applicationState.kafkaStreamsStates(false);
+
+        final Set<String> rackAwareAssignmentTags = new HashSet<>(getRackAwareAssignmentTags(applicationState));
+        final TagStatistics tagStatistics = new TagStatistics(applicationState);
+
+        final ConstrainedPrioritySet standbyTaskClientsByTaskLoad = standbyTaskPriorityListByLoad(streamStates, kafkaStreamsAssignments);
+
+        final Set<TaskId> statefulTaskIds = applicationState.allTasks().values().stream()
+            .filter(TaskInfo::isStateful)
+            .map(TaskInfo::id)
+            .collect(Collectors.toSet());
+        final Map<UUID, KafkaStreamsAssignment> clientsByUuid = kafkaStreamsAssignments.entrySet().stream().collect(Collectors.toMap(
+            entry -> entry.getKey().id(),
+            Map.Entry::getValue
+        ));
+        final Function<KafkaStreamsState, Map<String, String>> clientTagGetter = createClientTagGetter(applicationState);
+
+        final Map<TaskId, ProcessId> pendingStandbyTasksToClientId = new HashMap<>();
+        for (final TaskId statefulTaskId : statefulTaskIds) {
+            for (final KafkaStreamsAssignment assignment : clientsByUuid.values()) {
+                if (assignment.tasks().containsKey(statefulTaskId)) {
+                    assignStandbyTasksToClientsWithDifferentTags(
+                        numStandbyReplicas,
+                        standbyTaskClientsByTaskLoad,
+                        statefulTaskId,
+                        assignment.processId(),
+                        rackAwareAssignmentTags,
+                        streamStates,
+                        kafkaStreamsAssignments,
+                        tasksToRemainingStandbys,
+                        tagStatistics.tagKeyToValues,
+                        tagStatistics.tagEntryToClients,
+                        pendingStandbyTasksToClientId,
+                        clientTagGetter
+                    );
+                }
+            }
+        }
+
+        if (!tasksToRemainingStandbys.isEmpty()) {
+            assignPendingStandbyTasksToLeastLoadedClients(clientsByUuid,
+                numStandbyReplicas,
+                standbyTaskClientsByTaskLoad,
+                tasksToRemainingStandbys);
+        }
+
+        return kafkaStreamsAssignments;
+    }
+
+    private static Map<ProcessId, KafkaStreamsAssignment> loadBasedStandbyTaskAssignment(final ApplicationState applicationState,
+                                                                                         final Map<ProcessId, KafkaStreamsAssignment> kafkaStreamsAssignments) {
+        final int numStandbyReplicas = applicationState.assignmentConfigs().numStandbyReplicas();
+        final Map<TaskId, Integer> tasksToRemainingStandbys = applicationState.allTasks().values().stream()
+            .collect(Collectors.toMap(TaskInfo::id, taskInfo -> numStandbyReplicas));
```

Comment:
ditto here, move down a bit and then we can just initialize it as `statefulTaskIds.stream().collect(toMap(t -> t, t -> numStandbyReplicas))`